src/plugins/hyperbee/hyperbee.cpp

```diff
@@ -0,0 +1,1033 @@
+    json_object_object_add(jobj, "GFN",
+                           json_object_new_string_fmt("0x%"
+                                                      PRIx64, info->trap->memaccess.gfn));
+    fprintf(fp, "%s\n", json_object_get_string(jobj));
+    fclose(fp);
+
+    json_object_put(jobj);
+}
+
+/**
+ * Method to get the SHA256 hash of a memory area. Similar to the dump_memory_region-code in the memdump plugin.
+ */
+const gchar *getSHA256Memory(
+        vmi_instance_t vmi,
+        access_context_t *ctx,
+        size_t len_bytes) {
+    void **access_ptrs = nullptr;
+
+    const gchar *chk_str = nullptr;
+    GChecksum *checksum = nullptr;
+
+    addr_t aligned_addr;
+    addr_t intra_page_offset;
+    size_t aligned_len;
+    size_t len_remainder;
+    size_t num_pages;
+    size_t tmp_len_bytes = len_bytes;
+
+    aligned_addr = ctx->addr & ~(VMI_PS_4KB - 1);
+    intra_page_offset = ctx->addr & (VMI_PS_4KB - 1);
+
+    aligned_len = len_bytes & ~(VMI_PS_4KB - 1);
+    len_remainder = len_bytes & (VMI_PS_4KB - 1);
+
+    if (len_remainder) {
+        aligned_len += VMI_PS_4KB;
+    }
+
+    ctx->addr = aligned_addr;
+    num_pages = aligned_len / VMI_PS_4KB;
+
+    access_ptrs = (void **) g_malloc(num_pages * sizeof(void *));
+
+    if (VMI_SUCCESS != vmi_mmap_guest(vmi, ctx, num_pages, access_ptrs)) {
+        PRINT_DEBUG("[HYPERBEE] Failed mmap guest\n");
+        goto error;
+    }
+
+    checksum = g_checksum_new(G_CHECKSUM_SHA256);
+
+    for (size_t i = 0; i < num_pages; i++) {
+        // sometimes we are supposed to write less than the whole page
+        size_t write_length = tmp_len_bytes;
+
+        if (write_length > VMI_PS_4KB - intra_page_offset)
+            write_length = VMI_PS_4KB - intra_page_offset;
+
+        if (access_ptrs[i]) {
+            g_checksum_update(checksum, (const guchar *) access_ptrs[i] + intra_page_offset, write_length);
+            munmap(access_ptrs[i], VMI_PS_4KB);
+        } else {
+            // unaccessible page, pad with zeros to ensure proper alignment of the data
+            uint8_t zeros[VMI_PS_4KB] = {};
+            g_checksum_update(checksum, (const guchar *) zeros + intra_page_offset, write_length);
+        }
+
+        // this applies only to the first page
+        intra_page_offset = 0;
+        tmp_len_bytes -= write_length;
+    }
+
+    chk_str = g_checksum_get_string(checksum);
+    goto done;
+
+    error:
+    PRINT_DEBUG("[HYPERBEE] Failed to calculate checksum\n");
+
+    done:
+    g_free(access_ptrs);
+
+    return chk_str;
+}
+
+/**
+ * Dumps the memory specified by access context, from `ctx->addr` (first byte) to `ctx->addr + len_bytes - 1` (last byte).
+ * File is stored in file_path
+ *
+ * Similar to the dump_memory_region-code in the memdump plugin.
+ */
+bool
+dump_memory_region(vmi_instance_t vmi, hyperbee *plugin, access_context_t *ctx, size_t len_bytes, char *file_path) {
```

Comment:
I belive this part is copied from somewhere else. In such case it would be recommended to refactor the duplicate code into `libdrakvuf`. This is the place to keep common parts that are used across different plugins.

Reply:
This part is not copied, but it is based on some of the code of the memdump plugin (as refered in l. 309). Since I changed quite a bit (for example I refactored the method to calculate the hash at first to look it up in a map) I don't see a chance right now to make a common method. The only way would be to add more parameters (as options) but that would blow the method up and might be not that nice.

Reply:
I'm not opting for a strong refactor in cases where it would generate lots of trouble, of course. Do you identify any other regions in which the refactor into `libdrakvuf` would be +/- easy?

Reply:
What I can say is that the code I uploaded is more or less just part of my thesis with removing the private code parts. Through the many options it is rather complex. Another thing making this code as complex is the time intensive behaviour analysis (and my slow computer) so I optimized a lot regarding trying to prevent doing this analysis redundant. 

Three aspects:
A) I guess it would be possible with little effort to implement a way simpler plugin doing the same thing but less optimized.
We could discuss if the flags MALWARE_DEFAULT, DUMP_ALWAYS and LOG_ALWAYS are required or just could removed by agreeing to always dump and log. 

B) I would love to get your opinion on the modes of dumping the page or a vad node. I described in the main post the pros/cons I see in dumping vad.

C) Furthermore, If you have any ideas regarding the IGNORE_SYSTEM_DLL flag I would be interested in that as well.

Reply:
Ok, I can review this but not immediately. I hope this will be possible tomorrow :)

Reply:
There is absolutely no need to hurry. I'm very happy you are interested in what I did!

Reply:
> I'm not opting for a strong refactor in cases where it would generate lots of trouble, of course. Do you identify any other regions in which the refactor into `libdrakvuf` would be +/- easy?

I'll have a look.
- It might be interesting (especially for new drakvuf contributor) to have parametrized methods to create different trap types. It is a rather complex struct and one does not know what options should be chosen for the needed use. At least that was one of my problems, so I was lucky I found some code which was creating a exec_trap.
- My mm_access_fault_hook_cb method is very similar to the one in the ipt plugin. I just added this filter option. If we generalize this filter command line option we could reuse my extended method. Additionally I added lots of comments to document for myself what the mm_access_fault_hook_cb code is doing. They could be reused as well.
- The other thing is I saw many lines of the following form. I don't know if its worth to refactor this to a own method? Additionally I have no idea why asprintf could fail.
```
if (asprintf(&file_path_memory_dump, "%s/%s.vad", ef_data->plugin->hyperbee_dump_dir.c_str(),
                         file_name_stem) < 0) {
                PRINT_DEBUG("[HYPERBEE] Could not create memory dump file name\n");
            }
```

Reply:
> It might be interesting (especially for new drakvuf contributor) to have parametrized methods to create different trap types. It is a rather complex struct and one does not know what options should be chosen for the needed use. At least that was one of my problems, so I was lucky I found some code which was creating a exec_trap.

I think the trap API is already in the process of rework by @BonusPlay in #1115 , so this is subject to change soon.

Reply:
> DUMP_VAD: Lets you dump the whole VAD node instead of just the page. The problem I saw with it is, that the vad content (as collection of pages) is loaded dynamically on access. On the one hand hashing of vad nodes is more or less not reasonable since whenever a new page is loaded within the vad node the whole node is dumped and leads to another hash. On the other hand by being able to see which parts within the vad are loaded/accessed/executed after another might be interesting for malware analysis. The page mode is quicker, as it just has to hash 4096 bytes each time.

Yup, I was also struggling with this. I think this tactics is ought to work if the malware would extract itself into a new anonymous VAD - everything was recently written, so the pages will be accessible and VAD dump should be more-or-less complete. It's however problematic in more complex scenarios, as you've pointed out, when some pages are actually lazy-loaded.

> IGNORE_SYSTEM_DLL If we assume that the system's integrity is ensured turning on this flags discards the dump of all pages/vads whose vad name contains either System32 and SysWOW64. This gave me a major speedup since my PC is rather old and I was not interested in calls within the common Windows libraries/executables.

I think this is tricky as well. Malware might load some DLL and then patch it in memory. There is a copy-on-write mechanism in Windows that makes it pretty easy. Such writes will not be recognized as loss of integrity. Some programs even patch their own DLLs as a feature (e.g. to make sure that some API behaves consistently between different systems/versions). So we would possibly have to account for that as well.